### ISSUE TITLE
Compiling on linux: giving me a weird error

### ISSUE THREAD

**Issue by GuythatUhmCreatesRandomStuff** [djoslin0/sm64ex-coop#255]
https://github.com/djoslin0/sm64ex-coop/assets/144390627/8065904c-201c-4754-b1d9-5314588e5435

i am trying to compile sm64ex-coop on debian and it always errors out on this specific crash handler error, how do i fix this?

**Comment by robertkirkman** [comment]
This video is a recording of something that very closely resembles: **a Debian ARM chroot or proot**
**The `crash_handler.c` only supports `amd64` GNU/Linux.** It is unimplemented for other architectures and C libraries. To progress past that error, I recommend something like this:
```diff
diff --git a/src/pc/crash_handler.c b/src/pc/crash_handler.c
index d1e7d66d8..2c6f1ff88 100644
--- a/src/pc/crash_handler.c
+++ b/src/pc/crash_handler.c
@@ -3,7 +3,7 @@
 
 char gLastRemoteBhv[256] = "";
 
-#if (defined(_WIN32) || defined(__linux__)) && !defined(WAPI_DUMMY)
+#if defined(_WIN32) && !defined(WAPI_DUMMY)
 
 #ifdef HAVE_SDL2
 #include <SDL2/SDL.h>
```
If my assumptions about your environment are correct, you will encounter other challenges that require further code editing. There are solutions to every one, but should you choose to discuss them here further, **I recommend you move to a Raspberry Pi and post your logs from that, because it's the only ARM CPU device for which support has historically been merged into this repository.**

**Comment by GuythatUhmCreatesRandomStuff** [comment]
@robertkirkman does it mean i have to manually compile and install Rasb. pi OS ._.

**Comment by GuythatUhmCreatesRandomStuff** [comment]
since i use termux*

**Comment by robertkirkman** [comment]
No, what I mean by that is that I suspect **the only ARM CPU hardware device that this issues page is intended for is the Raspberry Pi hardware and other ARM hardware is not meant to be discussed here** .

I know that you are using Termux:X11, and I can also inform you that to my knowledge, I am the only person who has ever successfully played sm64ex-coop with GPU acceleration using Termux:X11, however, I don't think this page is meant for discussion of that topic.

Based on my assumptions about your environment, **I suspect you haven't installed any graphics drivers yet**, something that isn't connected to your error, but which you should be prepared for beforehand because **some devices supported by Termux:X11 are unfortunately unable to provide it _any_ access to their GPUs due to driver limitations**. In order to access your GPU, Termux:X11 requires manual setup of one of more graphics drivers as appropriate for the desired GPU. **Before trying further to use sm64ex-coop in your environment, I would suggest that you research which graphics driver you will need, set it up, and test it with tools like `glxgears` and `glmark2` to make sure your specific device will provide you with useful Termux:X11 GPU performance**

**Comment by GuythatUhmCreatesRandomStuff** [comment]
@robertkirkman sadly, my gpu drivers might be hard to find since it uses a very cheap gpu and well, its not from a popular branding or even a branding i have ever heard. It also has a very VERY confusing name.

**Comment by robertkirkman** [comment]
I am willing to help you connect Termux:X11 to your GPU and also help you install sm64ex-coop in whatever way you want, but I think this channel most likely isn't considered appropriate for the use of Termux:X11. You should go to the #x11 or #support channels on the official Termux Discord guild and ask further there - it's more appropriate for your situation than the termux-x11 repository's issues page, as you've seen, since you need support with solved problems rather than any unique bug in Termux:X11.

**Comment by GuythatUhmCreatesRandomStuff** [comment]
@robertkirkman oh ok.

**Comment by GuythatUhmCreatesRandomStuff** [comment]
@robertkirkman does rasbian lite(raspberry pi but lighter, like no pre-installed browser) work with sm64ex-coop?

**Comment by robertkirkman** [comment]
Yes it can but by default, there are errors like the one you posted in your first message. You would need to learn how to fix the errors, which I have done before, so it is possible.

**Comment by Isaac0-dev** [comment]
oh yeah that compile error will be my fault, I didn't test arm support for the linux crash handler, although I did guess it would happen
i added rpi support to coop a few updates ago but i havent really been maintaining it.. might be time to find my rpi again

**Comment by olafmustafar** [comment]
Hi, I think I can help with this, I am trying to build it on a RPI3 B, on Debian buster.
I made the changes and got the following error instead: 
![image](https://github.com/djoslin0/sm64ex-coop/assets/42854638/f9e9c76c-ef70-4442-935f-41c6ad2100be)
 
@robertkirkman @Isaac0-dev , any ideas how to proceed?

**Comment by Isaac0-dev** [comment]
coopnet is [open source](https://github.com/djoslin0/coopnet), you can try compiling it yourself
I haven't tried building coop on an arm device for a while so there may be more issues I'm unaware of. I'll try to do build on my rpi some time soon though to fix any issues that I might find.

**Comment by olafmustafar** [comment]
> can you describe how you did the second and third steps in detail

Sure!

## step 2
First I created a new folder `coopnet/lib/linux/arm`, then I moved the libjuice files there, like this:
![image](https://github.com/djoslin0/sm64ex-coop/assets/42854638/24e64d60-e674-4958-b30f-708d392007ac)

I also did some changes on the Makefile to (1) to reference those files when building for RPI and (2) build using c++20 standart, since the code is not c++11 compliant:
```diff --git a/Makefile b/Makefile
index 5f77e95..fe4d07e 100644
--- a/Makefile
+++ b/Makefile
@@ -4,11 +4,12 @@
 
 OSX_BUILD ?= 0
 LOGGING ?= 0
+TARGET_RPI ?= 0
 
 #################
 
 CXX = g++
-CXXFLAGS = -Wall -Werror -Wno-unused-function -std=c++11 -fPIC -DJUICE_STATIC -g
+CXXFLAGS = -Wall -Werror -Wno-unused-function -std=c++2a -fPIC -DJUICE_STATIC -g
 INCLUDES = -Icommon -Ilib/include
 LDFLAGS = -pthread
 SERVER_LIBS = -lcurl
@@ -41,6 +42,9 @@ else ifeq ($(OSX_BUILD),1)
   DYNLIB_NAME := libcoopnet.dylib
   LIBS := -l juice
   LDFLAGS += -rpath . -dynamiclib -install_name @rpath/$(DYNLIB_NAME)
+else ifeq($(TARGET_RPI),1)
+  CXXFLAGS += -Wno-nonnull-compare
+  LIB_DIR := lib/linux/arm
 else
   CXXFLAGS += -Wno-nonnull-compare
   LIB_DIR := lib/linux
```
after that I compiled with the new TARGET_RPI flag and it worked fine

## step 3
For step 3 I did basically the same, copied the files from libjuice and coopnet to it's own folder in the sm64 directory and changed the CMakeLists.txt, like this

```diff --git a/Makefile b/Makefile
index 2ef923aa..5d0f1398 100644
--- a/Makefile
+++ b/Makefile
@@ -912,7 +912,7 @@ ifeq ($(WINDOWS_BUILD),1)
     LDFLAGS += -mconsole
   endif
 else ifeq ($(TARGET_RPI),1)
-  LDFLAGS := $(OPT_FLAGS) -lm $(BACKEND_LDFLAGS) -no-pie
+  LDFLAGS := $(OPT_FLAGS) -lm $(BACKEND_LDFLAGS) -no-pie -lpthread
 else ifeq ($(OSX_BUILD),1)
   LDFLAGS := -lm $(BACKEND_LDFLAGS) -lpthread
 else
@@ -991,13 +991,15 @@ ifeq ($(COOPNET),1)
     ifneq (,$(findstring aarch64,$(machine)))
       LDFLAGS += -Llib/coopnet/linux -l:libcoopnet-arm64.a -l:libjuice.a
     else
-      LDFLAGS += -Llib/coopnet/linux -l:libcoopnet-arm.a -l:libjuice.a
+      LDFLAGS += -Llib/coopnet/linux/arm -l:libcoopnet.a -l:libjuice.a
     endif
   else
     LDFLAGS += -Llib/coopnet/linux -l:libcoopnet.a -l:libjuice.a
```
